devito/interfaces.py

```diff
@@ -1,22 +1,110 @@
 import numpy as np
-from sympy import IndexedBase, Symbol
+from sympy import IndexedBase, Function
 from tools import aligned
 
 
 __all__ = ['DenseData', 'TimeData', 'PointData']
 
 
-class DenseData(Symbol):
-    def __init__(self, name, shape, dtype):
-        self.name = name
-        self.shape = shape
-        self.dtype = dtype
-        self.pointer = None
-        self.initializer = None
-        super(DenseData, self).__init__(name)
+# This cache stores a reference to each created data object
+# so that we may re-create equivalent symbols during symbolic
+# manipulation with the correct shapes, pointers, etc.
+_SymbolCache = {}
 
-    def __new__(cls, name, shape, dtype):
-        return Symbol.__new__(cls, name)
+
+class SymbolicData(Function):
+    """Base class for data classes that provides symbolic behaviour.
+
+    :param name: Symbolic name to give to the resulting function. Must
+                 be given as keyword argument.
+    :param shape: Shape of the underlying spatial data. Must be given
+                  as keyword argument.
+
+    This class implements the symbolic behaviour of Devito's data
+    objects by inheriting from and mimicking the behaviour of :class
+    sympy.Function:. In order to maintain meta information across the
+    numerous re-instantiation SymPy performs during symbolic
+    manipulation, we inject the symbol name as the class name and
+    cache all created objects on that name. This entails that data
+    object should implement `__init__` in the following format:
+
+    def __init__(self, *args, **kwargs):
+        if self._cached():
+            SymbolicData.__init__(self)
+            return
+        else:
+            ... # Initialise object properties from kwargs
+
+            self._cache_put(self)
+
+    Note: The parameters :param name: and :param shape: must always be
+    present and given as keyword arguments, since SymPy uses `*args`
+    to (re-)create the dimension arguments of the symbolic function.
+    """
+
+    def __new__(cls, *args, **kwargs):
+        if cls not in _SymbolCache:
+            name = kwargs.get('name')
+            shape = kwargs.get('shape')
+            if len(args) < 1:
+                args = cls.indices(shape)
+            cls.__name__ = name  # Inject the given variable name
+        return Function.__new__(cls, *args)
+
+    def __init__(self):
+        """Initialise from a cached instance by shallow copying __dict__"""
+        original = _SymbolCache[self.__class__]
```

Comment:
No, because we (like SymPy itself) inject the symbol's name into the class, ie. `u = Function.__new__(cls, *args)`, where `cls.__name__ = 'u'`. The caching is then done on the class, which is now called `'u'`, so if you create two `DenseData(name='u')` they over-write in the cache, but `DenseData(name='v')` does not and creates a second entry.